### ISSUE TITLE
Training only a subset of synaptic weights

### ISSUE THREAD

**Issue by WI-KIWI** [tensorflow/tensorflow#6692]
It is often necessary to train only a subest of synaptic weights. One should be able to specify an arbitrary list of individual weights that will be updated/kept fixed by the gradient decent optimization algorithm and by other training algorithms. 

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

There are no solutions for training only a subset of individual synaptic weights. One is only able to define an entire tensor as not trainable, but it is not possible to define individual weights within a tensor as trainable / non-trainable.

**Comment by tillahoffmann** [comment]
You can pass a `var_list` to the [`minimize`](https://www.tensorflow.org/api_docs/python/train/optimizers#Optimizer.minimize) call of the optimizer.

**Comment by aselle** [comment]
@WI-KIWI  does @tillahoffmann's suggestion satisfy your need?

**Comment by szb0** [comment]
If I understand @WI-KIWI may be asking if you can apply a binary masking matrix or vector to some parameters at each step. I am also interested in functionality that would allow me to pass either a list of indices to zero out weights or a list of indices to keep active at each step.

**Comment by WI-KIWI** [comment]
Hi all,
First of all, thanks for your quick responses.

The suggestion of @thilahoffman works only with entire sensors, not with individual locations within a sensor. The latter is what I need.

@szb0 yes, an application of a binary mask would be the way to go. This makes it two of us who need this feature.

**Comment by michaelisard** [comment]
Sorry if I am not understanding, but can you just multiply by a mask as in the [dropout](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_ops.py#L1886) op?

**Comment by WI-KIWI** [comment]
I wouldn’t know what to multiply by a mask, so that the gradient descent would only train a needed subset of synapses (not tensors, but only certain elements within tensors). 

Of course, it is easy to define a control tensor of 1s and 0s indicating fixed and trainable weights. However, I don’t know how to make the gradient descent function use the control tensor.

**Comment by szb0** [comment]
Basically a way to easily apply binary masks to the gradients for each parameter matrix/vector. In particular a nice to have would be a way that allows for a sparse representation of the masks (non-zero or zero indices). I think this is already possible to do if you make a modification between the optimizers compute_gradients and apply_gradients call. Maybe just by defining placeholders for the indices at that will be zeroed out at each step and applying these masks to the gradients before applying them.

**Comment by tillahoffmann** [comment]
You could do something like [this](https://gist.github.com/tillahoffmann/c04dd6ca9259ef68f743e44c74e3d095).

**Comment by WI-KIWI** [comment]
@tillahoffmann thank you so much. I followed your suggestion and it worked! I appreciated your help!

Also many thanks to all here, especially to @szb0

Just WOW!  

Cheers

**Comment by jart** [comment]
@WI-KIWI Based on my cursory glance of this issue, it seems what @tillahoffmann suggested was able to solve your problem. In that case, a new feature might not be necessary. Therefore I'm going to close this issue. If you disagree, and still feel that a feature should be added, let me know and I'll re-open this issue.
